src/ir/ir_traversing_visitor_test.cc

```diff
@@ -132,25 +138,176 @@ TEST(IRTraversingVisitorTest, TraversesOperationAsExpected) {
                              std::make_unique<Module>());
   auto plus_op_module = plus_op_instance.impl_module();
 
-  CollectingVisitor preorder_visitor(CollectingVisitor::TraversalType::PRE);
+  CollectingVisitor preorder_visitor(CollectingVisitor::TraversalType::kPre);
   plus_op_instance.Accept(preorder_visitor);
   EXPECT_THAT(
       preorder_visitor.nodes(),
       testing::ElementsAre(std::addressof(plus_op_instance), plus_op_module));
 
-  CollectingVisitor postorder_visitor(CollectingVisitor::TraversalType::POST);
+  CollectingVisitor postorder_visitor(CollectingVisitor::TraversalType::kPost);
   plus_op_instance.Accept(postorder_visitor);
   EXPECT_THAT(
       postorder_visitor.nodes(),
       testing::ElementsAre(plus_op_module, std::addressof(plus_op_instance)));
 
-  CollectingVisitor all_order_visitor(CollectingVisitor::TraversalType::BOTH);
+  CollectingVisitor all_order_visitor(CollectingVisitor::TraversalType::kBoth);
   plus_op_instance.Accept(all_order_visitor);
   EXPECT_THAT(
       all_order_visitor.nodes(),
       testing::ElementsAre(std::addressof(plus_op_instance), plus_op_module,
                            plus_op_module, std::addressof(plus_op_instance)));
 }
+}  // namespace
+
+namespace {
+using ResultType=std::vector<void*>;
+
+// Another small test visitor that collects the nodes as they are visited to make
+// sure that  Pre and Post visits happened correctly.
+// Uses 'returns' rather than side effects.
+class ReturningVisitor : public IRTraversingVisitor<ReturningVisitor, ResultType> {
```

Comment:
Not a huge deal, but I wonder if there is a way to avoid code duplication here. The code and the tests are quite similar to `CollectingVisitor`.

Reply:
If we could template the tests on the visitor this could work. I've deduplicated what I could and will look into it.

Reply:
Ah, on second thought the visitors actually have different base types and different members. I don't know how to do this (Initially I thought we could have them both based on another class which exposed the nodes, but I don't think this works).